lib/config/global.ts

```diff
@@ -1,40 +1,56 @@
 import type { RenovateConfig, RepoGlobalConfig } from './types';
 
-let repoGlobalConfig: RepoGlobalConfig = {};
-
-// TODO: once global config work is complete, add a test to make sure this list includes all options with globalOnly=true (#9603)
-const repoGlobalOptions = [
-  'allowCustomCrateRegistries',
-  'allowPlugins',
-  'allowPostUpgradeCommandTemplating',
-  'allowScripts',
-  'allowedPostUpgradeCommands',
-  'binarySource',
-  'customEnvVariables',
-  'dockerChildPrefix',
-  'dockerImagePrefix',
-  'dockerUser',
-  'dryRun',
-  'exposeAllEnv',
-  'migratePresets',
-  'privateKey',
-  'privateKeyOld',
-  'localDir',
-  'cacheDir',
-];
-
-export function setGlobalConfig(
-  config: RenovateConfig | RepoGlobalConfig = {}
-): RenovateConfig {
-  repoGlobalConfig = {};
-  const result = { ...config };
-  for (const option of repoGlobalOptions) {
-    repoGlobalConfig[option] = config[option];
-    delete result[option];
+export class GlobalConfig {
+  // TODO: once global config work is complete, add a test to make sure this list includes all options with globalOnly=true (#9603)
+  private static readonly OPTIONS = [
+    'allowCustomCrateRegistries',
+    'allowedPostUpgradeCommands',
+    'allowPlugins',
+    'allowPostUpgradeCommandTemplating',
+    'allowScripts',
+    'binarySource',
+    'cacheDir',
+    'customEnvVariables',
+    'dockerChildPrefix',
+    'dockerImagePrefix',
+    'dockerUser',
+    'dryRun',
+    'exposeAllEnv',
+    'localDir',
+    'migratePresets',
+    'privateKey',
+    'privateKeyOld',
+  ];
+
+  private static config: RepoGlobalConfig = {};
+
+  static get(): RepoGlobalConfig;
+  static get<Key extends keyof RepoGlobalConfig>(
+    key?: Key
+  ): RepoGlobalConfig[Key];
+  static get<Key extends keyof RepoGlobalConfig>(
+    key?: Key
+  ): RepoGlobalConfig | RepoGlobalConfig[Key] {
+    return key ? GlobalConfig.config[key] : GlobalConfig.config;
   }
```

Comment:
Interesting! You can define the overloads, but you don't have to explicitly call them. I've not seen this pattern before :smile:

Reply:
Yeah, typescript will only use first two decalrations for caller type check and the last for local type check